corehq/ex-submodules/casexml/apps/phone/utils.py

```diff
@@ -48,21 +49,26 @@ def get_or_cache_global_fixture(restore_state, cache_bucket_prefix, fixture_name
     domain = restore_state.restore_user.domain
 
     data = None
+    key = '{}/{}'.format(cache_bucket_prefix, domain)
+
     if not restore_state.overwrite_cache:
         data = get_cached_fixture_items(domain, cache_bucket_prefix)
+        _record_datadog_metric('cache_miss' if data is None else 'cache_hit', key)
 
     if data is None:
-        with CriticalSection(['{}/{}'.format(cache_bucket_prefix, domain)]):
-            # re-check cache to avoid re-computing it
-            data = get_cached_fixture_items(domain, cache_bucket_prefix)
-            if data is not None:
-                return [data]
-            else:
-                items = data_fn()
-                io_data = write_fixture_items_to_io(items)
-                data = io_data.read()
-                io_data.seek(0)
-                cache_fixture_items_data(io_data, domain, fixture_name, cache_bucket_prefix)
+        with CriticalSection([key]):
```

Comment:
I'm guessing that right now, with the loss of the synclog database means that we have a ton of fixture updates that need to be sent. While looking for the code that prevents the sync I noticed it is actually looping through all products before sending instead of a SQL query https://github.com/dimagi/commcare-hq/blob/085e8aed4e253739aa381fde674c0106f364fd54/corehq/apps/commtrack/fixtures.py#L67-L71

Since we're seeing that this is constantly getting un-cached and we need to send full restores for users, I think that this `CriticalSection` means that we're getting 100 webworkers trying to run this code at the same time. it doesn't have to take one process 60 seconds if there's a ton of process waiting, then its the accumulation of all processes' time once the lock is acquired. (sorry if that's  confusing, I haven't quite figured out the best way to word this...)

Reply:
I'm not completely sure what you're saying but I agree this is inefficient. I didn't address this initially since this whole function was removed.

Reply:
I think what I was trying to say was if 100 restores come in at the same time and we assume that each loop in this critical section takes 1 second (for attempting to re-get the fixture from blob, retrieving the products, and pushing the new object to the blob storage) then for the first 60 seconds we would have these locks released every second and there would only be one process running at each time. On the 61st second the remaining 40 processes would enter the `CriticalSection` ([since fail_hard is defaulted to False](https://github.com/dimagi/commcare-hq/blob/e57c11388bd8d8616cdce481df29985c4ba1245b/corehq/ex-submodules/dimagi/utils/couch/__init__.py#L365)) so we would have a lot of processes trying to update the same record in blobmeta that we were seeing before.

Its not clear to me why they would get past the second object storage lookup at this time though

Reply:
that makes sense. Originally I had this as a non-blocking lock which returned and empty fixture if the lock failed to be acquired but for first syncs that would not work. Since we have they restore state in this function we could check if it's an initial sync or not and only block for initial syncs.